### ISSUE TITLE
🛠️ Unify error codes and add them into our translation checker

### ISSUE THREAD

**Issue by brindasasi** [GCTC-NTGC/gc-digital-talent#3330]
> [!IMPORTANT]
> This is an old ticket so it has been re-scoped based on discussions that took place in the comments. You can look there for more context.

## ♻️ Debt/Refactor

We have a mixed approach of how we deal with errors being sent via graphql. We would like to unify it and ensure we have translations for all error messages.

## 🕵️ Details

We recently added the ability to have enums be localized on the backend and it may make sense to register possible error codes as enum values that can be localized into a human readable error message. This would automatically have the in our server side intl check.

## 🙋‍♀️ Proposed Solution

Use a localized enum with a special handler to get a localized error message.

```php
enum ApiError {
  
  use HasLocalization;

  case EXPIRY_DATE_REQUIRED
  
  public function localizedErrorMessage()
  {
     $message = $this->localize();
     
     // Need to return a flat assoc for laravel messages
     return [ ...$message['value'], ...$mesage['label'] ];
  }
}
```

## 🌎 Localization

Most messages should already be localized but we may have missed one (hence the point of this ticket :sweat_smile: ) So you may need to have some translated.

## ✅ Acceptance Criteria

- [ ] Error code pattern is unified
- [ ] All codes are localized
- [ ] Missing srings are flagged in CI

---

## Original

**Description**
Given that we have implemented #3322/ #3328 for custom error messages .
But still we have to maintain those error codes in the front end manually in order to get extracted for translations.
It is easy to forget and not to add those api error codes in the front end. 
Hence we would like to extend intl-extract tool to smart enough to find the error code from graphql file as well 

Proposed solution:
Add logic to the translation workflow to check for enum in laravel and verify that the messages exist in the message file.

 Adding missing translations is OUT OF SCOPE

**A/C**
- [ ] intl-extract tool should find the messages from graphql as well and create the english json file 
- [ ] remove all the front end code that adds , uses the error message from the front end common messages file

**Comment by esizer** [comment]
Has this been solved with the updates to the error exchange?

https://github.com/GCTC-NTGC/gc-digital-talent/blob/417bf6edacbf3f374f70a4456e2291d5be4edeee/packages/client/src/components/ClientProvider/ClientProvider.tsx#L159-L181

**Comment by tristan-orourke** [comment]
I think this is still relevant - though I don't know how to implement it

**Comment by petertgiles** [comment]
I like the idea of hooking into the new localization stuff for the enums but I think that actually using enums for errors is a bit muddy.  Can we extend the tooling for use with errors using a similar technique without actually turning them into enums?  Without having to duplicate too much code?

**Comment by esizer** [comment]
> I like the idea of hooking into the new localization stuff for the enums but I think that actually using enums for errors is a bit muddy. Can we extend the tooling for use with errors using a similar technique without actually turning them into enums? Without having to duplicate too much code?

Yeah, I think if we start returning actual messages we could get rid of [the current enum](https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/api/database/helpers/ApiErrorEnums.php) and then the tooling shouldn't be too hard to tweak or extend.

We could either tweak the current method to accept a filename and keys instead of an enum. The main thing would be maintainign the list of expected keys for strings. Though, we could just use Englisgh as the source of truth? 🤔 

https://github.com/GCTC-NTGC/gc-digital-talent/blob/5867c74511b4c1d13342e942a0673bc31c974ec4/api/app/Console/Commands/CheckIntl.php#L107-L116

```php
class SomeValidator
{
  public function messages(): array {
     return [
        // This could also be a generic helper like `get_error_message($key)` 🤷‍♂ 
        'error_key' => [
            'en' => Lang::get('errors.error_key', [], 'en'),
            'fr' => Lang::get('errors.error_key', [], 'fr')
         ]
     ];
  }
}
```

**Comment by petertgiles** [comment]
I like this direction!  Yeah, if we don't use enums then the question is how to make sure we don't miss any.  :thinking:   Maybe you're right - don't over complicate it and just use an enum.  :shrug:

**Comment by esizer** [comment]
> I like this direction! Yeah, if we don't use enums then the question is how to make sure we don't miss any. 🤔 Maybe you're right - don't over complicate it and just use an enum. 🤷

Okay, so since you can only send back strings apparently (sort of makes sense). I have concocted something that could work. I wanted to get some thoughts before actually doing any work.

```php
// ApiErrorCodes.php
enum ApiErrorCodes
{
     // All the localization stuff..

    case EMAIL_UNIQUE;

    public function message(): string
    {
         return json_encode($this->localize());    
    }
}

// UpdateUserInputValidator.php
class UpdateUserInputValidator
{
   public function messages(): array
   {
       return [
           'email.unique' => ApiErrorCodes::EMAIL_UNIQUE->message()
       ];
    }
}
```



![Image](https://github.com/user-attachments/assets/0453c496-daa1-4116-b78b-e3e694cafb06)

**Comment by petertgiles** [comment]
I like it!  Two quick comments...

On this part:
```php
public function message(): string
    {
         return json_encode($this->localize());    
    }
```

It seems weird to see "message(): string" and then see it returning a json_encode.  Can we call the function "localizedMessage" or "errorResponse" or "jsonMessage" or something more specific?

On this part:
```
 public function messages(): array
   {
       return [
           'email.unique' => ApiErrorCodes::EMAIL_UNIQUE->message()
       ];
    }
```

If `ApiErrorCodes::EMAIL_UNIQUE->message()` returns an array instead of an already encoded json string, will Laravel automatically encode it for you when it sends?  It would be nice and php-y to not have to manually json_encode in the enum and instead return a regular array.  It would solve my first comment, too.

**Comment by esizer** [comment]
> It seems weird to see "message(): string" and then see it returning a json_encode. Can we call the function "localizedMessage" or "errorResponse" or "jsonMessage" or something more specific?

Yeah, that makes sense. I like `jsonMessage` since it is very explicit about what it should be :+1:  

> If `ApiErrorCodes::EMAIL_UNIQUE->message()` returns an array instead of an already encoded json string, will Laravel automatically encode it for you when it sends? It would be nice and php-y to not have to manually json_encode in the enum and instead return a regular array. It would solve my first comment, too.

okay so oddly enough I got this working with just `['en' => 'EN', 'fr' => 'FR']` but it stopped working with the proper localized value :cry: 

So it seems like it must return an array of strings and won't auto encode :-1: I'm not sure why it initially worked because laravel calls:

https://github.com/laravel/framework/blob/81a0b5b60ff83d7a47177675741c41c1dfe78967/src/Illuminate/Validation/Validator.php#L403

Which works only on strings, unless maybe something in the actual localized array was giving PHP casting issues but I was getting an "array to string conversion" error.

tldr, we must ensure the message is a string.

**Comment by esizer** [comment]
This is going super poorly with custom rules 😑 I am also not super happy with it the further In get. You can check out my draft PR that is honestly a mess. It works in some ways we do validation but not others and I am struggling to make it compatible across the board.

The massive point of contention is that all the tooling around this expects strings that are translatable, not arrays that contain the translated values.

It works when we do our own unique way of just throwing out the errors ourselves but then we can't use the normal test methods to test errors because it expects a proper validation message be returned. But it falls apart when we use the provided ways of validating and allow laravel/lighthouse-php handle the response.

I have come to the conclusion that this will not work 100% so we probably should not pursue it.

## Where do we go from here?

I have two additional ideas but it could get a little hairy 😅 

### Make laravel aware of the users preferred language

Right now laravel has no idea what language the content is being displayed in which requires us to respond with both possible languages. This is fine and works well so no reason to change this. However, if we did we could return simple strings to the user for error messages. How could we do this?

There are a few options we could explore but honestly I see a clear winner so no need to go into them unless someone disagrees.

 - GraphQL Arg - Bad because it requires us to add it to all our mutations/queries that could error
 - Cookie - Not great but could work and will get sent with every request and stored, but not great because it gets stored
 - Header - I think this is the best option, gets sent withe very request and then we don't need to store it

We could update our client to be aware of the current locale (it currently sits inside of the locale provider so we can get it from context). Then we could add a grpahql middleware that reads the header and sets the current locale. Once laravel is aware of the locale, we can use their built in translation to send just the validation string already translated.

This would also could possible reduce the number of messages we need because then we have access to replacement values.

```tsx
// ClientProvider.tsx
const locale = getLocale(intl);

createClient({
  fetchOptions: { 
     headers: { 'Accept-Language': locale, }
 }
})
```

```php
class LocaleMiddleware
{
   public function handle(Request $request, Closure $next)
   {
      $locale = $request()->headers->get('Accept-Language') ?? 'en';
      App::setLocale($locale);

     $next($request);
  }
}

class ApiErrorEnum
{
    case SOME_ERROR_CODE;

    public function localzedMessage(?array $replacementValues): string
    {
        return __($this->getLangKey(), $replacementValues);
    }
}

class SomeValidationRule {
       public function validate(string $attribute, mixed $value, Closure $fail): void
       {
           // Logic
           $fail(ApiError::SOME_CODE->message(['attribute' => $attribute]);
       }
}

// api_error.php
return [
   'some_error_code' => 'Validation failed for :attribute'
];
```

**Comment by petertgiles** [comment]
I like this in concept, I'd love to see how this looks when you start coding it.  Header makes sense to me, seems like I see this pattern a lot with APIs.

How would this interact all enum work you did?  Could we cut down on overfetching and just return enum label in the requested language?

This could be nice for server-side generated files and emails, too.  We could use the language the user is in rather than relying on them to keep the profile up to date.

**Comment by esizer** [comment]
> How would this interact all enum work you did? Could we cut down on overfetching and just return enum label in the requested language?

I was kind of thinking about that and even how it works with not just the enums but the `LocalizedString` type as well if we really wanted to. Though, that gets tricky because sometimes we do want both locales (admin editing). 

For localized enums, since we only care about the `key => value` we could do this:

```graphql
type LocalizedEnum {
   value: String!;
   label: String!;
}
```

```php
$resolver = function ($value, $args, $context, ResolveInfo $info) use ($enum) {
    switch ($info->fieldName) {
        case 'value': return $value;
        case 'label': return $enum::localize($value);
        default: return null;
   }
);

trait HasLocalization {
    public static function localize(string $value)
    {
        $key = self::getLangKey($value);
        
        return Lang::get($key, []);
    }
}
```

> This could be nice for server-side generated files and emails, too. We could use the language the user is in rather than relying on them to keep the profile up to date.

I didn;t event think of that but it is a nice side effect!

**Comment by esizer** [comment]
And looks to work :+1: This is going to be a lot of work so I think I'll create some tickets to break this work up. I'm thinking:

The issue I am having now is naming, we already have `localizedString` which is exactly what I'd call this :sweat_smile: 

- Make requests locale aware
- Update localized enums return single value
- Move error codes to localized enum
- Refactor localized string to return current locale alongside localized string (optional/low priority)

@tristan-orourke what do you think about the proposed tickets?

![2024-08-22_10-10](https://github.com/user-attachments/assets/02424a33-b2d1-4cbd-a014-d27e6b239f19)
![2024-08-22_10-10_1](https://github.com/user-attachments/assets/505ac10c-0f9e-4264-94d7-596f4e59b4a9)
